### PR TITLE
add rocket chat client

### PR DESCRIPTION
**Pull request by exmike**
Made on the basis of the old pull request - https://github.com/qa-guru/allure-notifications/pull/163
Only with integration rocket chat
Result: 
image

### CODE REVIEW

File: src/main/java/guru/qa/allure/notifications/Application.java [hunk 1]
```diff
@@ -3,7 +3,6 @@
 import guru.qa.allure.notifications.clients.Notification;
 import guru.qa.allure.notifications.config.ApplicationConfig;
 import guru.qa.allure.notifications.config.Config;
-import guru.qa.allure.notifications.exceptions.MessagingException;
```
Comment (valfirst) [inline src/main/java/guru/qa/allure/notifications/Application.java:6]: please, make sure do avoid adding unrelated changes

File: src/main/java/guru/qa/allure/notifications/clients/ClientFactory.java [hunk 2]
```diff
@@ -18,7 +19,7 @@ public class ClientFactory {
     public static List<Notifier> from(Config config) {
         MessageData messageData = new MessageData(config.getBase());
 
-        List<Notifier> notifiers =  new ArrayList<>();
+        List<Notifier> notifiers = new ArrayList<>();
```
Comment (valfirst) [inline src/main/java/guru/qa/allure/notifications/clients/ClientFactory.java:22]: please, make sure do avoid adding unrelated changes

File: src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import guru.qa.allure.notifications.config.rocket.Rocket;
+import guru.qa.allure.notifications.exceptions.MessagingException;
+import guru.qa.allure.notifications.json.JSON;
+import guru.qa.allure.notifications.template.RocketTemplate;
+import guru.qa.allure.notifications.template.data.MessageData;
+import java.io.ByteArrayInputStream;
+import kong.unirest.ContentType;
+import kong.unirest.Unirest;
+
+public class RocketClient implements Notifier {
```
Comment (valfirst) [inline src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java:13]:
```suggestion
public class RocketChatClient implements Notifier {
```

File: src/main/resources/templates/rocket.ftl [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    **-${environment}:** ${env}
+    **-${comment}:** ${comm}
+    **-${totalScenarios}:** ${total}
+    <#if passed != 0 > **-${totalPassed}:** ${passed} </#if>
+    <#if failed != 0 > **-${totalFailed}:** ${failed} </#if>
+    <#if broken != 0 > **-${totalBroken}:** ${broken} </#if>
+    <#if unknown != 0 > **-${totalUnknown}:** ${unknown} </#if>
+    <#if skipped != 0 > **-${totalSkipped}:** ${skipped} </#if>
+    <#if reportLink??>**${reportAvailableAtLink}:** ${reportLink}</#if>
+</#compress>
```
Comment (valfirst) [inline src/main/resources/templates/rocket.ftl:12]: is there any reason to have additional template?

File: src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    public RocketClient(MessageData messageData, Rocket rocket) {
+        this.rocket = rocket;
+        this.template = new RocketTemplate(messageData);
+    }
+
+    @Override
+    public void sendText() throws MessagingException {
+        String body = String.format("{\"channel\": \"%s\", \"text\": \"%s\" }",
+            rocket.getChannel(), template.create().replace("\r\n", "\\\n"));
```
Comment (valfirst) [inline src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java:27]: why not let `Unirest` take care of JSON body creation?

File: src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        String url = String.format("%s/api/v1/chat.postMessage", rocket.getUrl());
```
Comment (valfirst) [inline src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java:28]:
```suggestion
        String url = rocket.getUrl() + "/api/v1/chat.postMessage";
```

Comment (valfirst) [inline src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java:28]: what if `rocket.getUrl()` ends with `/`?

File: src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        Unirest.post(url)
+            .header("X-Auth-Token", rocket.getToken())
+            .header("X-User-Id", rocket.getUserId())
+            .header("Content-Type", "application/json")
```
Comment (valfirst) [inline src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java:32]:
```suggestion
            .header("Content-Type", ContentType.APPLICATION_JSON.getMimeType())
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: src/main/resources/templates/rocket.ftl (hunk 4 above)
Comment (exmike) [reply src/main/resources/templates/rocket.ftl:12]: with default template have some issues 
image

File: src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java (hunk 6 above)
Comment (exmike) [reply src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java]: it works fine
image

File: src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java (hunk 5 above)
Comment (exmike) [reply src/main/java/guru/qa/allure/notifications/clients/rocket/RocketClient.java]: ok, rewrited

File: src/main/java/guru/qa/allure/notifications/clients/rocket/RocketChatClient.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import guru.qa.allure.notifications.template.data.MessageData;
+import java.io.ByteArrayInputStream;
+import java.util.HashMap;
+import java.util.Map;
+import kong.unirest.ContentType;
+import kong.unirest.Unirest;
+
+public class RocketChatClient implements Notifier {
+
+    private final Map<String, Object> body = new HashMap<>();
```
Comment (valfirst) [inline src/main/java/guru/qa/allure/notifications/clients/rocket/RocketChatClient.java:16]: it should be a variable in method `sendText`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: src/main/java/guru/qa/allure/notifications/clients/rocket/RocketChatClient.java (hunk 8 above)
Comment (exmike) [reply src/main/java/guru/qa/allure/notifications/clients/rocket/RocketChatClient.java]: ok, fixed

**Review by valfirst — APPROVED** [review]
